File: odyssey-api/src/routes/account.ts
import { Hono } from "hono";
import { eq } from "drizzle-orm";
import { users } from "../db/schema";
import type { Env } from "../types";
import { createDb } from "../db/client";

const app = new Hono<{ Bindings: Env; Variables: { userId: string } }>();

// DELETE /account — Delete all user data (GDPR cascade delete)
app.delete("/", async (c) => {
  const userId = c.get("userId");
  const db = createDb(c.env.DATABASE_URL);

  // CASCADE on FK will delete entries, key_backups, sync_log
  await db.delete(users).where(eq(users.clerkUserId, userId));

  return c.json({ deleted: true });
});

export default app;

File: deprecated/odyssey-server/src/routes/account.rs
use axum::extract::State;
use axum::routing::delete;
use axum::{Json, Router};
use serde::Serialize;

use crate::AppState;
use crate::UserScope;
use crate::auth::AuthUser;
use crate::error::AppError;

pub fn router() -> Router<AppState> {
    Router::new().route("/account", delete(delete_account))
}

#[derive(Debug, Serialize)]
pub struct DeleteResponse {
    pub deleted: bool,
}

async fn delete_account(
    State(state): State<AppState>,
    user: AuthUser,
) -> Result<Json<DeleteResponse>, AppError> {
    let scope = UserScope::new(&user, &state);
    // ON DELETE CASCADE handles entries, encrypted_key_backups, sync_log
    sqlx::query("DELETE FROM users WHERE clerk_user_id = $1")
        .bind(&scope.user_id)
        .execute(&scope.pool)
        .await?;

    Ok(Json(DeleteResponse { deleted: true }))
}
